constantine/signatures/ecdsa.nim

```diff
@@ -308,4 +301,110 @@ func coreVerify*[Pubkey, Sig](
+
+proc recoverPubkeyImpl_vartime*[Name: static Algebra; Sig](
+    recovered: var EC_ShortW_Aff[Fp[Name], G1],
+    signature: Sig,
+    msgHash: Fr[Name],
+    evenY: bool) =
+  ## Attempts to recover an associated public key to the given `signature` and
+  ## hash of a message `msgHash`.
+  ##
+  ## Note that as the signature is only dependent on the `x` coordinate of the
+  ## curve point `R`, two public keys verify the signature. The one with even
+  ## and the one with odd `y` coordinate (one even & one odd due to curve prime
+  ## order).
+  ##
+  ## `evenY` decides whether we recover the public key associated with the even
+  ## `y` coordinate of `R` or the odd one. Both verify the (message, signature)
+  ## pair.
+  ##
+  ## If the signature is invalid, `recovered` will be set to the neutral element.
+  type
+    ECAff = EC_ShortW_Aff[Fp[Name], G1]
+    ECJac = EC_ShortW_Jac[Fp[Name], G1]
+  # 1. Set to neutral so if we don't find a valid signature, return neutral
+  recovered.setNeutral()
+  const G = Name.getGenerator($G1)
+
+  let rInit = signature.r.toBig() # initial `r`
+  var x1 = Fp[Name].fromBig(signature.r.toBig()) # as coordinate in Fp
+  let M = Fp[Name].fromBig(Fr[Name].getModulus())
+
+  # Due to the conversion of the `x` coordinate in `Fp` of the point `R` in the signing process
+  # to a scalar in `Fr`, we potentially reduce it modulo the subgroup order (if `x > M` with
+  # `M` the subgroup order).
+  # As we don't know if this is the case, we need to loop until we either find a valid signature,
+  # adding `M` each iteration or until we roll over again, in which case the signature is invalid.
+  # NOTE: For secp256k1 this is _extremely_ unlikely, because prime of the curve `p` and subgroup
+  # order `M` are so close!
+  var validSig = false
+  while (not validSig) and bool(x1.toBig() <= rInit):
+    # 1. Get base `R` point
+    var R {.noinit.}: EC_ShortW_Aff[Fp[Name], G1]
+    let valid = R.trySetFromCoordX(x1) # from `r = x1`
+    if not bool(valid):
+      x1 += M # add modulus of `Fr`. As long as we don't overflow in `Fp` we try again
+      continue # try next `i` in `x1 = r + i·M`
+
+    let isEven = R.y.toBig().isEven()
+    # 2. only negate `y ↦ -y` if current and target even-ness disagree
+    R.y.cneg(isEven xor SecretBool evenY)
+
+    # 3. perform recovery calculation, `Q = -m·r⁻¹ * G + s·r⁻¹ * R`
+    # Note: Calculate with `r⁻¹` included in each coefficient to avoid 3rd `scalarMul`.
+    var rInv = signature.r
+    rInv.inv() # `r⁻¹`
+
+    var u1 {.noinit.}, u2 {.noinit.}: Fr[Name]
+    u1.prod(msgHash, rInv)     # `u₁ = m·r⁻¹`
+    u1.neg()                   # `u₁ = -m·r⁻¹`
+    u2.prod(signature.s, rInv) # `u₂ = s·r⁻¹`
+
+    var Q {.noinit.}: ECJac # the potential public key
+    var point1 {.noinit.}, point2 {.noinit.}: ECJac
+    point1.scalarMul(u1, G)    # `p₁ = u₁ * G`
+    point2.scalarMul(u2, R)    # `p₂ = u₂ * R`
+    Q.sum(point1, point2)      # `Q = p₁ + p₂`
```

Comment:
This comes up quite often for ECDSA and should be a dedicated proc, unsure about the name yet `dualScalarMul`, `simul`, `lincomb`.

The multiscalarmul_vartime proc has too much overhead for just 2:
![image](https://github.com/user-attachments/assets/442e5edb-5e85-4edf-aa11-d55a7c5560f4)
but there are the Strauss or the Yao algorithms to accelerate this, especially given that it's a bottleneck for TLS:
- https://github.com/mratsim/constantine/issues/36
- https://cr.yp.to/papers/pippenger-20020118-retypeset20220327.pdf

Reply:
Also the `scalarMulEndo` routines like this https://github.com/mratsim/constantine/blob/6b65b0e77a4ea3435b7ba2a7a9e01aaf704653a5/constantine/math/elliptic/ec_scalar_mul_vartime.nim#L252-L331

actually are implemented in 2 steps:
- splitting a scalar in 2 or 4
- and computing a dual or quadruple scalar multiplications efficiently, including in constant-time.

So it should be possible to isolate those internals in a separate PR and redirect the publicly exposed `dualScalarMul` to use these and minimize extra code that needs to be written.

Reply:
Issue for it here: https://github.com/mratsim/constantine/issues/507